### PR TITLE
impl(storage): support upload hashing in `RawClient`

### PR DESCRIPTION
**Pull request by coryan**
This changes the `storage::internal::HashFunction` classes to support calling them from `RawClient::InsertObjectMedia()` and `RawClient::UploadChunk()`. The new functionality is not used yet, but it is testable and thus can be merged as a separate PR.

Part of the work for #11060

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11065)

### CODE REVIEW

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 1]
```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
...
+  EXPECT_THAT(a2.crc32c, IsEmpty());
+  EXPECT_THAT(a2.md5, kQuickFoxMD5Hash);
+}
+
+TEST(HashFunctionImplTest, MD5StringViewWithCrc) {
+  MD5HashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    auto const message_crc = storage_internal::Crc32c(message);
```
Comment (dbolduc) [inline google/cloud/storage/internal/hash_function_impl_test.cc:173]: this just gets ignored, right? I assume the MD5 and CRC hashes do not mix.

Might it be clearer to say something like: `std::uint32_t const ignored_crc = 0`; ?

(If the real implementation provides a crc32c to the MD5 `Update` function, then I am okay with the test as is.)

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 2]
```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
...
                              absl::make_unique<Crc32cHashFunction>());
-  Update(function, "The quick");
-  Update(function, " brown");
-  Update(function, " fox jumps over the lazy dog");
+  function.Update("The quick");
+  function.Update(" brown");
+  function.Update(" fox jumps over the lazy dog");
   auto result = std::move(function).Finish();
   EXPECT_THAT(result.crc32c, kQuickFoxCrc32cChecksum);
   EXPECT_THAT(result.md5, kQuickFoxMD5Hash);
```
Comment (dbolduc) [inline google/cloud/storage/internal/hash_function_impl_test.cc:222]: should we call `Finish()` twice and see if `ubsan` hollers?

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 3]
```diff
@@ -196,23 +308,43 @@ TEST(HasFunctionImplTest, CreateHashFunctionUpload) {
...
+
+TEST(HashFunctionImplTest, CreateHashFunctionInsertObjectMedia) {
+  struct Test {
+    DisableCrc32cChecksum crc32c;
+    DisableMD5Hash md5;
+    HashValues expected;
+  } const cases[] = {
+      {DisableCrc32cChecksum(false), DisableMD5Hash(false),
+       HashValues{kQuickFoxCrc32cChecksum, kQuickFoxMD5Hash}},
+  };
```
Comment (dbolduc) [inline google/cloud/storage/internal/hash_function_impl_test.cc:338]: Did you want to include these 9200 test cases?

https://github.com/googleapis/google-cloud-cpp/blob/74926358837d73d05b8157a138f280cb71e39573/google/cloud/storage/internal/hash_function_impl_test.cc#L260-L304

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 4]
```diff
@@ -196,23 +308,43 @@ TEST(HasFunctionImplTest, CreateHashFunctionUpload) {
         ResumableUploadRequest("test-bucket", "test-object")
             .set_multiple_options(test.crc32_disabled, test.crc32_value,
                                   test.md5_disabled, test.md5_value));
-    Update(*function, "The quick brown fox jumps over the lazy dog");
+    function->Update("The quick brown fox jumps over the lazy dog");
```
Comment (dbolduc) [inline google/cloud/storage/internal/hash_function_impl_test.cc:311]: nit: can use `kQuickFox`

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 5]
```diff
@@ -130,14 +242,14 @@ TEST(HasFunctionImplTest, CreateHashFunctionRead) {
     auto function = CreateHashFunction(
         ReadObjectRangeRequest("test-bucket", "test-object")
             .set_multiple_options(test.crc32_disabled, test.md5_disabled));
-    Update(*function, "The quick brown fox jumps over the lazy dog");
+    function->Update("The quick brown fox jumps over the lazy dog");
```
Comment (dbolduc) [inline google/cloud/storage/internal/hash_function_impl_test.cc:245]: nit: can use `kQuickFox`

**Review by dbolduc — CHANGES_REQUESTED** [review]
I could be wrong, but it is worth checking the inequality in `AlreadyHashed(...)`

File: google/cloud/storage/internal/hash_function_impl_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc]: Fixed.

File: google/cloud/storage/internal/hash_function_impl_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc]: Done.

File: google/cloud/storage/internal/hash_function_impl_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc]: Done.

File: google/cloud/storage/internal/hash_function_impl_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc]: renamed to `unused`.

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 6]
```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
...
+TEST(HashFunctionImplTest, Crc32cStringView) {
+  Crc32cHashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    EXPECT_STATUS_OK(function.Update(npos, message));
+    EXPECT_THAT(function.Update(npos, message),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc]: Done.

File: google/cloud/storage/internal/hash_function_impl.cc [hunk 7]
```diff
@@ -46,30 +51,116 @@ void DeleteDigestCtx(EVP_MD_CTX* context) {
 #endif  // OPENSSL_VERSION_NUMBER >= 0x10100000L
 }
 
+template <typename Buffer>
+bool AlreadyHashed(std::int64_t offset, Buffer const& buffer,
+                   std::int64_t minimum_offset) {
+  auto const end = offset + buffer.size();
+  return static_cast<std::int64_t>(end) < minimum_offset;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl.cc]: Yup.

File: google/cloud/storage/internal/hash_function_impl_test.cc [hunk 8]
```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
...
 
-TEST(HasFunctionImplTest, EmptyMD5) {
+TEST(HashFunctionImplTest, Crc32cStringView) {
+  Crc32cHashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    EXPECT_STATUS_OK(function.Update(npos, message));
+    EXPECT_THAT(function.Update(npos, message),
+                StatusIs(StatusCode::kInvalidArgument));
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl_test.cc:85]: Fixed.

File: google/cloud/storage/internal/hash_function_impl.cc [hunk 9]
```diff
@@ -46,30 +51,116 @@ void DeleteDigestCtx(EVP_MD_CTX* context) {
...
+
+Status MD5HashFunction::Update(std::int64_t offset, absl::string_view buffer,
+                               std::uint32_t /*buffer_crc*/) {
+  if (offset == minimum_offset_) {
+    Update(buffer);
+    minimum_offset_ += buffer.size();
+    return {};
+  }
+  if (AlreadyHashed(offset, buffer, minimum_offset_)) return {};
+  return InvalidArgumentError("mismatched offset", GCP_ERROR_INFO());
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl.cc]: Done.

File: google/cloud/storage/internal/hash_function_impl.cc [hunk 10]
```diff
@@ -46,30 +51,116 @@ void DeleteDigestCtx(EVP_MD_CTX* context) {
...
+Status CompositeFunction::Update(std::int64_t offset, absl::Cord const& buffer,
+                                 std::uint32_t buffer_crc) {
+  auto status = a_->Update(offset, buffer, buffer_crc);
+  if (!status.ok()) return status;
+  return b_->Update(offset, buffer, buffer_crc);
 }
 
-HashValues CompositeFunction::Finish() && {
+HashValues CompositeFunction::Finish() {
   return Merge(std::move(*a_).Finish(), std::move(*b_).Finish());
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl.cc]: No, fixed.

File: google/cloud/storage/internal/hash_function_impl.h [hunk 11]
```diff
@@ -17,9 +17,8 @@
 
 #include "google/cloud/storage/internal/hash_function.h"
 #include "google/cloud/storage/version.h"
+#include "absl/types/optional.h"
 #include <openssl/evp.h>
-#include <cstdint>
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/hash_function_impl.h:21]: Ooops

**Review by dbolduc — APPROVED** [review]

**Pull request merged**